sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -39,15 +63,17 @@ trait DateTimeFormatterHelper {
       .withResolverStyle(ResolverStyle.STRICT)
   }
 
-  protected def toInstantWithZoneId(temporalAccessor: TemporalAccessor, zoneId: ZoneId): Instant = {
-    val localTime = if (temporalAccessor.query(TemporalQueries.localTime) == null) {
-      LocalTime.ofNanoOfDay(0)
-    } else {
-      LocalTime.from(temporalAccessor)
+  def getFormatter(pattern: String, locale: Locale): DateTimeFormatter = {
+    try {
+      cache.get(
+        (pattern, locale),
+        new Callable[DateTimeFormatter]() {
```

Comment:
Not sure it matters but this makes a callable on every call. Java classes don't have any lazy eval. It could be saved and reused.

Reply:
I had to construct the lambda function in previous version too: https://github.com/apache/spark/pull/23462/commits/f08c71e2f3d61bbe9d70e528a6a022b776fafa1a#diff-3e4b85f54d2e75f1b5b845279749024eR45 This `_ => buildFormatter(pattern, locale)` creates new instance, correct?

Reply:
That also creates something that implement Callable, yes, and is a little cleaner, but I think the issue is that it gets instantiated every time? hardly a big deal, just wonder if this is moderately performance-sensitive.

Reply:
I'm not sure if there is a better way. To save the computation we must pass a "function" to `getOrElse`, and "function" in Java/Scala is an object. Another way is to sacrifice multi-thread contention a little bit:
```
synchronized {
  val formatter = cache.get...
  if (formatter == null) {
    val f = create formatter
    cache.put(key, f)
    f
  } else {
    formatter
  }
```

Reply:
Nevermind, I think there isn't a better way, yes. I thought this could be defined once as a function but even that would have to produce a lambda.

Reply:
If we suppose that patterns (and locale) are no changed frequently, and they present in cache most likely, I think it makes sense to avoid creation of `Callable` instance. If you don't mind, I will change the code to:
```scala
val key = (pattern, locale)
var formatter = cache.getIfPresent(key)
if (formatter == null) {
  formatter = buildFormatter(pattern, locale)
  cache.put(key, formatter)
}
formatter
```
without `synchronized {...}`. In rare cases, the formatter will be created twice but I think it is ok.